packages/headless-react/src/ssr/client-utils.ts

```diff
@@ -1,23 +1,30 @@
 'use client';
 
-import {useEffect, useReducer, useRef} from 'react';
+import {useEffect, useMemo, useReducer} from 'react';
+
+/**
+ * Subscriber is a function that takes a single argument, which is another function `listener` that returns `void`. The Subscriber function itself returns another function that can be used to unsubscribe the `listener`.
+ */
+export type Subscriber = (listener: () => void) => () => void;
+
+export type SnapshotGetter<T> = () => T;
 
 /**
  * Alternate for `useSyncExternalStore` which runs into infinite loops when hooks are used in `getSnapshot`
  * https://github.com/facebook/react/issues/24529
  */
 export function useSyncMemoizedStore<T>(
-  subscribe: (listener: () => void) => () => void,
-  getSnapshot: () => T
+  subscribe: Subscriber,
+  getSnapshot: SnapshotGetter<T>
 ) {
-  const stateRef = useRef(getSnapshot());
+  let latestSnapshot = useMemo(() => getSnapshot(), [getSnapshot, subscribe]);
   const [, forceRender] = useReducer((s) => s + 1, 0);
 
   useEffect(() => {
     let isMounted = true;
     const unsubscribe = subscribe(() => {
       if (isMounted) {
-        stateRef.current = getSnapshot();
+        latestSnapshot = getSnapshot();
```

Comment:
Is this update needed? I would expect this line of code to do nothing, since `latestSnapshot` will be reinitialized on the next render.

I think a test may be missing, something like "when the listener is called, re-renders and updates the snapshot". I suspect it would fail on this test. It's probably the reason why this fails in the e2e tests:
<img width="458" alt="image" src="https://github.com/coveo/ui-kit/assets/54454747/8fde40d0-2490-4b0f-932d-36c9efa02c88">

# Solutions

There's probably better solutions than those two, but they are the only ones I can really think of...

## Solution A: `useState`
One potential fix I thought of for this issue would be to replace `latestSnapshot`'s `useMemo` with a `useState`. 

However, unlike memos, states don't have a dependencies array. This means that, when something causes a re-render where `subscribe` and/or `getSnapshot` are updated, the state wouldn't get updated.

There may be ways to work around this issue. For example, you could update the state inside `useEffect`. However, this would cause a second re-render, which is pretty inefficient.

This is why I would fall back to solution B or C.

## Solution B: `useRef`

You could replace `latestSnapshot`'s `useMemo` with a [`useRef`](https://react.dev/reference/react/useRef).

However, similarly to `useState`, `useRef` doesn't have a dependencies array, it would need to be manually updated. Unlike `useState`, updating it wouldn't cause a re-renders and would be effective immediately. Beware though, updating it inside `useEffect` wouldn't affect the current render, since `useEffect` is called after rendering.

You could use the custom [`useHasChanged`](https://github.com/coveo/ui-kit/blob/KIT-1456-btaillon-app-router/packages/samples/headless-next/src/hooks/use-controller.ts#L122) hook I made last time to update it in the current rendering phase, but that's one more hook on the pile of hooks 😅 .

Reply:
Thanks! Switched to `useRef` with useHasChanged.
Missed the e2e test failure as I was re-triggering to get rid of some other flaky tests 🤦